### PR TITLE
Implement CLI (closes #105)

### LINKED ISSUE #105: Implement command line interface
So, I would like to start bikeshedding regarding CLI. Here is how I see it:
```
 Usage: testcafe [command] [options]


  Commands:

    run [config-name]            run tests with given configuration and/or options
    add-config <config-name>     create run configuration for given options
    remove-config <config-name>  remove run configuration

  If [command] is not specified "run" command will be used as default

  Options:

    -h, --help                     output usage information
    -V, --version                  output the version number
    -lb, --list-browsers           output available browsers
    -f, --files <file1>,...        files or glob patterns to run
    -b, --browsers <browser1>,...  browsers to run tests in
    -r, --reporter <name>          specify the reporter to use
    -rp, --report-path <path>      output report to the file on the <path>
    -s, --screenshots <path>       enable screenshots and save them to the <path>
    -sf, --screenshots-on-fails    take screenshots on test fails
    -g, --grep <pattern>           only run tests matching <pattern>
    -f, --fgrep <string>           only run tests containing <string>
    -fg, --fixture-grep <pattern>  only run fixtures matching <pattern>
    -ff, --fixture-fgrep <string>  only run fixtures containing <string>
    -nc, --no-colors               force disabling of colors
    -gc, --global-config           use global configuration file
```

Regarding configurations: the idea is what you can save commonly used configuration with some alias, e.g.:
```
> testcafe add-config ie-grid -b ie -f test/**/*.js -fg ASPxGridView
```

Then this configuration will be saved to the `.testcaferc` file in the root of the project.
Optionally you can use `-gc` flag. In that case configuration will be saved to the global configuration file and can be reused in other projects.
It's no required for the configuration to have all options, so you can adopt it for different projects by passing additional options on run:
```
> testcafe add-config ie-chrome -b ie,chrome
> testcafe ie-chrome -f test/**/*.js
``` 

What bothers me:
- Is this configuration approach clear
- Is this usage is clear enough (we can add link to the detailed docs, so it's kinda not a big deal)
- Due to console limitations lists should not contains spaces, just `,`. E.g.:

*this is correct:*
```
testcafe -b chrome,ie,firefiox
``` 
*this is not:*
```
testcafe -b chrome, ie, firefiox
``` 
I see it as a possible source for the user confusion.

Please, let me know what do you guys think and how we can make it better.
\cc @AlexanderMoskovkin @churkin @kirovboris @miherlosev @AlexanderMoskovkin @AndreyBelym @helen-dikareva @VasilyStrelyaev @MargaritaLoseva @arubtsov

### PR DESCRIPTION
**Pull request by inikulin**
You can try it locally (please try it) by running in the shell:
```sh
mkdir test-cli && cd ./test-cli && \
git clone https://github.com/inikulin/testcafe-phoenix && \ 
cd ./testcafe-phoenix && git checkout cli && \
npm install && gulp build
```

Then to run CLI:
```
node bin/testcafe <args>
```

\cc @churkin @AlexanderMoskovkin @kirovboris 
@VasilyStrelyaev check texts and test names, please

### CODE REVIEW

File: src/cli/index.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        .join('\n');
+
+    console.error(msg);
+    console.error(chalk.gray('Type "tescafe -h" for help.'));
+
+    exit(1);
+}
+
+function waitBrowserConnectionReady (browserConnection) {
+    return new Promise((resolve)=> browserConnection.once('ready', resolve));
```
Comment (kirovboris) [inline src/cli/index.js]: ```(resolve)``` -> ```resolve``` ?

Comment (inikulin) [inline src/cli/index.js]: good catch

File: src/cli/index.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+(async function cli () {
+    try {
+        var argParser = new CliArgumentParser();
+
+        await argParser.parse(process.argv);
+
+        if (argParser.opts.listBrowsers)
+            listBrowsers();
+
```
Comment (kirovboris) [inline src/cli/index.js]: Why are we making empty line here?

File: src/cli/index.js (hunk 1 above)
Comment (inikulin) [inline src/cli/index.js]: I'll add #108 as the part of this PR

File: src/client/test-run/iframe.js.mustache [hunk 3]
```diff
@@ -71,7 +71,7 @@
 
         testCafeCore.SETTINGS.set({
             TAKE_SCREENSHOT_ON_FAILS: {{{takeScreenshotOnFails}}},
-            FAIL_ON_JS_ERRORS:        {{{failOnJsErrors}}},
+            SKIP_JS_ERRORS:        {{{skipJsErrors}}},
```
Comment (kirovboris) [inline src/client/test-run/iframe.js.mustache]: Indentation is broken

File: src/client/test-run/index.js.mustache [hunk 4]
```diff
@@ -42,7 +42,7 @@
             CURRENT_TEST_STEP_NAME:   nextStep ? stepNames[nextStep - 1] : 'Test initialization',
             BROWSER_STATUS_URL:       '{{{browserStatusUrl}}}',
             TAKE_SCREENSHOT_ON_FAILS: {{{takeScreenshotOnFails}}},
-            FAIL_ON_JS_ERRORS:        {{{failOnJsErrors}}},
+            SKIP_JS_ERRORS:        {{{failOnJsErrors}}},
```
Comment (kirovboris) [inline src/client/test-run/index.js.mustache]: Indentation is broken

File: src/cli/index.js (hunk 2 above)
Comment (inikulin) [inline src/cli/index.js]: Actually we always use this style, for better readability. Unfortunately, we don't have es-rule to enforce it

Comment (inikulin) [inline src/cli/index.js]: However, looking at the source now, I see different usages, so we don't have consensus on this.

**Comment by kirovboris** [comment]
I'm executing ```node bin/testcafe chrome,ie test.js```, and getting 

```ERROR Cannot read property 'isTTY' of null```

**Comment by inikulin** [comment]
+ Hammerhead errors are not handled correctly

File: src/cli/argument-parser.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        }
+    }
+
+    async parse (argv) {
+        this.program.parse(argv);
+
+        this.opts = this.program.opts();
+
+        // NOTE: '-list-browsers' option just lists browsers and immediately exits app.
+        // Therefore, we don't need process other args
```
Comment (VasilyStrelyaev) [inline src/cli/argument-parser.js]: NOTE: the '-list-browsers' option only lists browsers and immediately exits the app.
Therefore, we don't need to process other arguments.

File: src/cli/index.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import chalk from 'chalk';
+import { getInstallations as getBrowserInstallations } from 'testcafe-browser-natives';
+import CliArgumentParser from './argument-parser';
+import createTestCafe from '../';
+
+const ERR_PREFIX = chalk.red('ERROR ');
+
+function exit (code) {
+    // NOTE: give process time to flush output.
+    // It's necessary for some environments.
```
Comment (VasilyStrelyaev) [inline src/cli/index.js]: NOTE: give a process time to flush the output.
It's necessary in some environments.

File: src/cli/index.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+function waitBrowserConnectionReady (browserConnection) {
+    return new Promise((resolve)=> browserConnection.once('ready', resolve));
+}
+
+async function createRemoteBrowserConnections (testCafe, remoteCount, showQRCode) {
+    var connections = [];
+
+    // NOTE: add blank line
```
Comment (VasilyStrelyaev) [inline src/cli/index.js]: NOTE: add a blank line

File: src/messages.js [hunk 8]
```diff
@@ -3,10 +3,15 @@ export const MESSAGE = {
...
+    browserNotSet:                      'No browser selected to test against.',
+    testSourcesNotSet:                  'No test file specified.',
     noTestsToRun:                       'No tests to run. Either the test files contain no tests or the filter function is too restrictive.',
-    cantFindReporterForAlias:           'The provided "{name}" reporter does not exist. Check that you have specified the report format correctly.'
+    cantFindReporterForAlias:           'The provided "{name}" reporter does not exist. Check that you have specified the report format correctly.',
+    optionValueIsNotValidRegExp:        '"{optionName}" option value is not valid regular expression.',
+    portNumberIsNotInteger:             'Port number should be a valid integer.',
+    portsOptionRequiresTwoNumbers:      '"--ports" option requires two numbers to be specified.',
+    portIsNotFree:                      'Specified {portNum} port is already in use by other program.',
+    invalidHostname:                    'Specified "{hostname}" hostname can not be resolved to the current machine.'
```
Comment (VasilyStrelyaev) [inline src/messages.js]:
```
optionValueIsNotValidRegExp:        'The "{optionName}" option value is not a valid regular expression.',
portNumberIsNotInteger:             'A port number should be a valid integer.',
portsOptionRequiresTwoNumbers:      'The "--ports" option requires two numbers to be specified.',
portIsOccupied:                     'The specified {portNum} port is already in use by another program.',
invalidHostname:                    'The specified "{hostname}" hostname cannot be resolved to the current machine.'
```

File: src/cli/argument-parser.js [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+            .option('-S, --screenshots-on-fails', 'take screenshots on test fails')
+            .option('-q, --quarantine-mode', 'enable quarantine mode')
+            .option('-e, --skip-js-errors', 'don\'t fail tests on JS errors on page')
+            .option('-t, --test <name>', 'only run tests with <name>')
+            .option('-T, --test-grep <pattern>', 'only run tests matching <pattern>')
+            .option('-f, --fixture <name>', 'only run fixtures with <name>')
+            .option('-F, --fixture-grep <pattern>', 'only run fixtures matching <pattern>')
+            .option('--ports <port1,port2>', 'specify custom port numbers')
+            .option('--hostname <name>', 'specify hostname')
+            .option('--qrcode', 'show QR-code for remote connections');
```
Comment (VasilyStrelyaev) [inline src/cli/argument-parser.js]: the correctors need some more time

Comment (VasilyStrelyaev) [inline src/cli/argument-parser.js]:
```
In the browser list, you can use aliases (e.g. "ie9", "chrome", etc.) as well as paths to executables.
To run tests against all installed browsers, use the "all" alias.
To use a remote browser connection (e.g. to connect a mobile device), specify "remote" as a browser alias. If you need to connect multiple devices, prefix the alias with the number of browsers you want to connect (e.g., "3remote").

You can use one or more file paths or glob patterns to specify which tests to run.

More info: http://testcafe.devexpress.com/Documentation/CLI

-b, --list-browsers				output the available browser aliases
-r, --reporter <name>			specify the reporter type to use
-p, --report-path <path>		specify the path to which the report file should be saved
-s, --screenshots <path>		enable taking screenshots and specify the path to save them to
-S, --screenshots-on-fails		take a screenshot whenever a test fails
-q, --quarantine-mode			enable the quarantine mode
-e, --skip-js-errors			make tests not fail when a JS error happens on a page 
-t, --test <name>				run only tests with the specified name  
-T, --test-grep <pattern>		run only tests matching the specified pattern  
-f, --fixture <name>			run only fixtures with the specified name 
-F, --fixture-grep <pattern>	run only fixtures matching the specified pattern 
--ports <port1,port2>			specify custom port numbers
--hostname <name>				specify the hostname
--qrcode						display QR-code you can use to easily establish remote connections
```

**Comment by VasilyStrelyaev** [comment]
\r-

File: test/server/cli-argument-parser-test.js [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+              '!test/server/data/file-list/dir2/file-2-1.js ' +
+              'test/server/data/file-list/dir3')
+            .then(function (parser) {
+                expect(parser.src).eql(expected);
+                done();
+            })
+            .catch(done);
+    });
+
+    it("Should parse screenshots path and ensure it's existence", function (done) {
```
Comment (VasilyStrelyaev) [inline test/server/cli-argument-parser-test.js]: one more - `its`

Comment (VasilyStrelyaev) [inline test/server/cli-argument-parser-test.js]: or even better
`Should parse the screenshot path and ensure it exists`

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by AlexanderMoskovkin** [comment]
Is this an indents issue or it's ok?

**Comment by inikulin** [comment]
@AlexanderMoskovkin Definitely no. Need to figure out optimal string lengths with the final texts (which we now have thanks to @VasilyStrelyaev)

File: src/client/test-run/index.js.mustache [hunk 11]
```diff
@@ -42,7 +42,7 @@
             CURRENT_TEST_STEP_NAME:   nextStep ? stepNames[nextStep - 1] : 'Test initialization',
             BROWSER_STATUS_URL:       '{{{browserStatusUrl}}}',
             TAKE_SCREENSHOT_ON_FAILS: {{{takeScreenshotOnFails}}},
-            FAIL_ON_JS_ERRORS:        {{{failOnJsErrors}}},
+            SKIP_JS_ERRORS:           {{{failOnJsErrors}}},
```
Comment (AlexanderMoskovkin) [inline src/client/test-run/index.js.mustache]: ```{{{failOnJsErrors}}}``` should be changed with ```{{{skipJsErrors}}}``` here

File: src/cli/index.js [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+import { getInstallations as getBrowserInstallations } from 'testcafe-browser-natives';
+import CliArgumentParser from './argument-parser';
+import createTestCafe from '../';
+
+const ERR_PREFIX = chalk.red('ERROR ');
+
+function exit (code) {
+    // NOTE: give process time to flush output.
+    // It's necessary for some environments.
+    setTimeout(() => process.exit(code), 0);
```
Comment (AlexanderMoskovkin) [inline src/cli/index.js]: We should increase the timeout to close browsers after the run. It works with about 1000 ms on my machine. It's a time to go to the idle page and close the browser.
TestCafe uses the broken vesion of `testcafe-browser-natives`, so to test this I've put the working `testcafe-browser-natives` module version to `node_modules`

Comment (inikulin) [inline src/cli/index.js]: You mean I need to bump version in `package.json`? 

> We should increase the timeout to close browsers after the run.

Sounds bad. Seems like it should be done on runner level and not CLI. Ideally, we shouldn't stop the task until all browser `close()` methods not stop working.

Comment (inikulin) [inline src/cli/index.js]: BTW, can you issue a separate ticket for this, please?

Comment (AlexanderMoskovkin) [inline src/cli/index.js]: >You mean I need to bump version in package.json?
The fixed version is not published yet

>Ideally, we shouldn't stop the task until all browser close() methods not stop working.
Agree. I'll create a separate ticket

Comment (AlexanderMoskovkin) [inline src/cli/index.js]: https://github.com/superroma/testcafe-phoenix/issues/123

File: src/client/test-run/index.js.mustache (hunk 11 above)
Comment (inikulin) [inline src/client/test-run/index.js.mustache]: Yes!

**Comment by inikulin** [comment]
FPR

**Comment by AlexanderMoskovkin** [comment]
lgtm

it works nice on my side

**Comment by kirovboris** [comment]
lgtm

**Comment by inikulin** [comment]
ping @churkin

**Comment by churkin** [comment]
lgtm

**Comment by churkin** [comment]
Waiting for @VasilyStrelyaev

File: src/cli/argument-parser.js [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+        this.program
+
+            .version(JSON.parse(read('../../package.json')).version)
+            .usage('[options] <comma-separated-browser-list> <file-or-glob ...>')
+            .description(CliArgumentParser._getDescription())
+
+            .option('-b, --list-browsers', 'output the available browser aliases')
+            .option('-r, --reporter <name>', 'specify the reporter type to use')
+            .option('-p, --report-path <path>', 'specify the path to which the report file should be saved')
+            .option('-s, --screenshots <path>', 'enable screenshot capturing and specify the path to save them to')
```
Comment (VasilyStrelyaev) [inline src/cli/argument-parser.js]: oops... 'them' doesn't relate ro anything anymore...
`enable screenshot capturing and specify the path to save the screenshots to`

**Comment by VasilyStrelyaev** [comment]
lgtm

**Pull request merged**